datafusion/physical-plan/src/sorts/sort.rs

```diff
@@ -1126,14 +1127,20 @@ impl ExecutionPlan for SortExec {
                 Ok(Box::pin(RecordBatchStreamAdapter::new(
                     self.schema(),
                     futures::stream::once(async move {
-                        while let Some(batch) = input.next().await {
-                            let batch = batch?;
-                            topk.insert_batch(batch)?;
-                            if topk.finished {
-                                break;
+                        // Spawn a task the first time the stream is polled for the sort phase.
+                        // This ensures the consumer of the sort does not poll unnecessarily
+                        // while the sort is ongoing
```

Comment:
I may not fully understand this change, but I believe it will start generating input as soon as `execute` is called, rather than waiting for the first poll... 

Another potential problem is that it will disconnect the production of batches from the consumption of them -- in other words by design I think it will potentially produce batches on a different thread than consumes them

Reply:
I've added a test case that attempts to demonstrate that processing is deferred. If this looks ok to you I can add the same thing for the other touched code as well.

I'm not sure how I can demonstrate the absence of multi-threading in a test case.

Wrt comprehensibility, I have to admit I still very much in the learning-as-I-go phase of using the futures crate. There might be a more elegant or straightforward way to express this construct.